uamqp/authentication.py

```diff
@@ -0,0 +1,177 @@
+#-------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+#-------------------------------------------------------------------------
+
+import time
+import urllib
+import hmac
+import hashlib
+import base64
+from collections import namedtuple
+from functools import partial
+
+from .sasl import SASLAnonymousCredential, SASLPlainCredential
+
+from .constants import (
+    AUTH_DEFAULT_EXPIRATION_SECONDS,
+    TOKEN_TYPE_JWT,
+    TOKEN_TYPE_SASTOKEN
+)
+
+try:
+    from urlparse import urlparse
+    from urllib import quote_plus  # type: ignore
+except ImportError:
+    from urllib.parse import urlparse, quote_plus
+
+AccessToken = namedtuple("AccessToken", ["token", "expires_on"])
+
+
+def _generate_sas_token(auth_uri, sas_name, sas_key, expiry_in=AUTH_DEFAULT_EXPIRATION_SECONDS):
+    auth_uri = urllib.parse.quote_plus(auth_uri)
+    sas = sas_key.encode("utf-8")
+    expires_on = int(time.time() + expiry_in)
+    string_to_sign = (auth_uri + '\n' + str(expires_on)).encode('utf-8')
+    signed_hmac_sha256 = hmac.HMAC(sas, string_to_sign, hashlib.sha256)
+    signature = urllib.parse.quote(base64.b64encode(signed_hmac_sha256.digest()))
+    return AccessToken(
+        "SharedAccessSignature sr={}&sig={}&se={}&skn={}".format(auth_uri, signature, str(expires_on), sas_name),
+        expires_on
+    )
+
+
+class SASLPlainAuth(object):
+    def __init__(self, authcid, passwd, authzid=None):
+        self.sasl = SASLPlainCredential(authcid, passwd, authzid)
+
+
+class _CBSAuth(object):
```

Comment:
Can we add a string literal discriminator for these `Auth` objects?

Reply:
emmm, wondering what is a "string literal discriminator", a class variable?

Reply:
Yes - exactly that. Just a string class variable:
```python
class Foo:
    kind = 'foo'
    value = 42

class Bar:
    kind = 'bar'
    data = 'some bytes'
```
That way, we can make a cleaner switch statement, rather than type-checking everything:
```python
if obj.kind == 'foo':  # can also use enums here
    output = obj.value
elif obj.kind == 'bar':
    output = obj.data
else:
    raise Exception("what's this")
```